pkg/endpoint/endpoint.go

```diff
@@ -1753,11 +1753,10 @@ func (e *Endpoint) metadataResolver(ctx context.Context,
 	pod, k8sMetadata, err := resolveMetadata(ns, podName)
 	switch {
```

Comment:
this switch/case statement is awkward to read; Now that you've got the complete flow in your head, would it be easy to just convert it to two `if` statements?

I _think_ so, but it's hard to tell, which is a good sign we should refactor this code.

Reply:
> this switch/case statement is awkward to read;

Personally agree, that's what misled me when writing the previous patch.

> would it be easy to just convert it to two if statements?

IMO the cleanest approach would be moving the UID check responsibility directly to `resolveMetadata`, so that the caller can just treat it as a plain error. That would also ensure that it is always validated by all callers. The change looks reasonably straightforward, although there are a few implications that would need to be carefully considered. For instance, the calls part of the `UpdateNoTrackRules` and `UpdateBandwidthPolicy` callbacks could then potentially return an error if there's an UID mismatch, and AFAICT there's no retry mechanism in place. My understanding is that these calls could be dropped altogether in favor of accessing the pod object just retrieved here, but I don't have deep enough knowledge of the endpoint subsystem to be 100% sure of that. 

I'd propose to keep the current patch minimal, and then tackle this refactoring as a follow-up, which doesn't need to be backported (I can open a PR to start the discussion). WDYT?

Reply:
Seems reasonable. The various edge cases of the endpoint state machine are unfortunate, to say the least.